packages/actor-rdf-resolve-hypermedia-links-queue-wrapper-info-occupancy/lib/LinkQueueLogger.ts

```diff
@@ -0,0 +1,164 @@
+    if (link.producedByActor &&
+      this.linkProductionRatio[event][link.producedByActor?.name]) {
+      this.linkProductionRatio[event][link.producedByActor?.name] += 1;
+    } else if (
+      this.linkProductionRatio[event][link.producedByActor?.name ?? 'unknown'] === undefined) {
+      this.linkProductionRatio[event][link.producedByActor?.name ?? 'unknown'] = 1;
+    } else {
+      this.linkProductionRatio[event].unknown += 1;
+    }
+  }
+
+  /**
+   * Create the information of a link queue event
+   * @param {ILink} link - current link
+   * @param {EventType} eventType - the type of event
+   * @param {ILink|undefined} parent - the parent of the link
+   * @returns {ILinkQueueEvent} current event of the link queue
+   */
+  private createLinkQueueEvent(link: ILink, eventType: 'PUSH' | 'POP', parent?: ILink): ILinkQueueEvent {
+    const linkInfo: IUrlStatistic = {
+      url: link.url,
+      producedByActor: LinkQueueLogger.getActorProductorInformation(link),
+      timestamp: performance.now(),
+      parent: parent?.url,
+    };
+    this.updateLinkProductionRatio(linkInfo, eventType === 'POP' ? 'popEvents' : 'pushEvents');
+
+    return {
+      type: eventType,
+      link: linkInfo,
+      query: this.query,
+      queue: {
+        size: this.getSize(),
+        ...this.linkProductionRatio,
+      },
+    };
+  }
+
+  public override push(link: ILink, parent: ILink): boolean {
+    const resp: boolean = super.push(link, parent);
+    if (resp) {
+      const event: ILinkQueueEvent = this.createLinkQueueEvent(link, 'PUSH', parent);
+      this.materialize(event);
+    }
+    return resp;
+  }
+
+  public override pop(): ILink | undefined {
+    const link = super.pop();
+    if (link !== undefined) {
+      const event: ILinkQueueEvent = this.createLinkQueueEvent(link, 'POP');
+      this.materialize(event);
+    }
+    return link;
+  }
+
+  /**
+   * Materialize the current event into the logger
+   * @param {ILinkQueueEvent} event - Current event
+   */
+  private materialize(event: ILinkQueueEvent): void {
+    const jsonEvent = { ...event, type: event.type };
+    this.logger.trace(LinkQueueLogger.LINK_QUEUE_EVENT_NAME, { data: JSON.stringify(jsonEvent) });
+  }
+}
+
+/**
+ * A link queue event
+ */
+interface ILinkQueueEvent {
+  type: 'PUSH' | 'POP';
+  link: IUrlStatistic;
+  query: string;
+  queue: IQueueStatistics;
+}
+/**
+ * Statistic of the link queue
+ */
+interface IQueueStatistics extends ILinkProductionActorRatio {
+  size: number;
+}
+/**
+ * Ratio of the actor producing links in relation to the push and pop events.
+ * The key of the index is the name of the actor and the value is the number of occurences.
+ */
+interface ILinkProductionActorRatio {
+  pushEvents: Record<string, number>;
+  popEvents: Record<string, number>;
+}
+
+/**
+ * Information about an URL
+ */
+interface IUrlStatistic {
+  url: string;
+  producedByActor: IProducedByActor | null;
```

Comment:
```suggestion
  producedByActor?: IProducedByActor;
```

Reply:
Avoid `null`!

Reply:
I want a null because they the field remain when it is converted to a JSON and the purpose of this object is to be turned into a string.

Reply:
But what's wrong with the field just not being there in that case?

Reply:
I like having it there :) Also I think it would be better to have a fixed schema. But it is more a personal thing. I can change it if it is wrong or it is too different from the rest of Comunica coding style.

Reply:
It's a TypeScript convention to prefer undefined over null: https://github.com/Microsoft/TypeScript/wiki/Coding-guidelines#null-and-undefined